lib/pure/times.nim

```diff
@@ -1121,39 +1207,161 @@ proc evaluateStaticInterval(interval: TimeInterval): Duration =
     minutes = interval.minutes,
     hours = interval.hours)
 
+proc between*(startDt, endDt:DateTime): TimeInterval =
+  ## Evaluate difference between two dates in ``TimeInterval`` format, so, it
+  ## will be relative.
+  ##
+  ## **Warning:** It's not recommended to use ``between`` for ``DateTime's`` in 
+  ## different ``TimeZone's``.  
```

Comment:
Is it ever useful to do so? If not then it should just be disallowed outright.

Reply:
Yes. and i still doubt the issues @GULPF  is raising here are valid. Whats the problem with e.g. asking how old you are if your birthdate is in timezone a and you are currently living in timezone b?

It is a well known fact that arithmetic with relative date/time differences does not work the same way as arithmetic with numbers. The most evident fact is that you usually get different results if you calculate between(a, b) and between(b, a) and the difference is not only the sign. Another example is that if you calculate between(a, b) you may get 17 years, 11 months and 27 days, and one second later you may get 18 years and not the expected 17 years, 11 months and 28 days (it only looks familiar if you start from december 31, 23:59:59). This has all nothing to do with timezones or DST switches, its a consequence of the relativity of the calculated results. They depend on the point in time you take as a reference to  perform the calculation. 

And then it is somewhat unclear what the best implementation of addition/subtraction of relative TimeIntervals might be. The current solution in times.nim which gives 2018-03-03 if you subtract 1 month from 2018-03-31 gives a hint on the problematic. This is clearly not the solution most other date/time libraries have choosen. And convincing someone that the time between 2018-02-28 and 2018-03-31 is really 1 month and 3 days is not the easiest task.

Reply:
> Whats the problem with e.g. asking how old you are if your birthdate is in timezone a and you are currently living in timezone b?

Because it's the wrong question. When you ask for your age, what you want to know is the calendar difference between the two dates. To calculate that **you should ignore the timezone**, not convert the DateTime to UTC. That's like asking "what would my age be if I was born in UTC and I currently live in UTC?", which is a pointless question if the person doesn't actually live in UTC. This is why I suggested implementing `between` for `NaiveDateTime` instead (which afaik is how other datetime libs does it).

If the two input timezones to `between` differs a lot, the result might be total nonsense. E.g the difference between `25/03 2010` and `24/03 2011` could be more than a year, which is not how most people think when they count their age. I don't see how such a result is useful.

> The current solution in times.nim which gives 2018-03-03 if you subtract 1 month from 2018-03-31 gives a hint on the problematic. This is clearly not the solution most other date/time libraries have choosen. And convincing someone that the time between 2018-02-28 and 2018-03-31 is really 1 month and 3 days is not the easiest task.

fwiw I don't like the current behavior either, but it's been that way for a long time. But I know at least one other language which uses the same logic by default: D.

Reply:
> If the two input timezones to between differs a lot, the result might be total nonsense. E.g the difference between 25/03 2010 and 24/03 2011 could be more than a year,

Thats true for eg. Kiritimati (UTC+14) and Midway Islands (UTC-11). Why is that total nonsense?

I know that many people do not adjust their age, if they dislocate to another timezone. But that is more an interesting cultural habit, than a problem due to the calculation of durations between two points in time in different timezones. 

www.timeanddate.com has this: https://www.timeanddate.com/date/timezoneduration.html
Are they calculating nonsense?

If you make a phone call from Kiritimati to someone in Honolulu and ask what date and time it is there, will you be surprised that the clock shows the same time but the calendar shows a day earlier? Whats the difficuty in calculating the duration of 2018-05-04T19:00:00 Kirimati Time and 2018-05-03T19:02:00 Honolulu Time.

Navigators in the past calculated their longitudinal position from the difference between Greenwich time and their local time as observed from celestial bodies (eg. local noon). This technique was rather successful and not pure nonsense, 

If you start a travel in Afghanistan at 21:00 and after 2 hours you reach the Chinese border. Do you have any problem with the fact that you already crossed the date-line, because the clocks of the chinese border control show 02:30 a day later? Why should it not be allowed to calculdate the duration between 2018-05-04T02:30 Chinese Time and 2018-05-03T21:00 Afghan Time. What if you lost your clock on the travel? Will you be unable to figure out how long you needed to reach the chinese border?

I really think @survivorm's `between` in its current form is a welcome and useful addition to times.nim. There is no need to "enhance" it by putting some limitations on the allowed timezones.

Reply:
> I know that many people do not adjust their age, if they dislocate to another timezone. But that is more an interesting cultural habit, than a problem due to the calculation of durations between two points in time in different timezones.

Well you mentioned it as a use case for `between`, so I demonstrated why it can't be used for that.

> I really think @survivorm's between in its current form is a welcome and useful addition to times.nim. There is no need to "enhance" it by putting some limitations on the allowed timezones.

I agree that adding such a limitation wouldn't improve the proc as it's currently written. My original argument was that `a + between(a, b) == b` could be guaranteed only if such a limitation was added, but since that's not desired anyway it's fine to allow different timezones.

My issue with `between` is that it counts calendar units in UTC, but all other operations with `TimeInterval` uses calendar units in the specified timezone. The fact that `between` returns UTC units seems arbitrary to me, and I don't see the benefit (especially not since the current behaviour could easily be achieved by doing `between(a.utc, b.utc)`).

Reply:
>  since the current behaviour could easily be achieved by doing between(a.utc, b.utc).

The point is: it is very hard to implement `between(a, b)` without converting `a` and `b`  first to a common base which conventionally is utc. If you are able to provide such an implementaion, that would also be very welcome. 

For something like scheduling (eg. marking a meeting taking place every friday in the last week of a month at 13:00 in an agenda) it is true that ignoring the timezone helps, but in the general case you can't simply ignore timezones and use timezone-unaware datetimes to calculate meaningful durations in calendrical units. The duration between 2018-05-04 in Honolulu and 2018-05-05 in Kiritimati is 0 days. And in São Paulo to calculate `between(2018-02-17T23:30:00, 2018-02-18T00:30:00)` you need to know that the DST ends at 2018-02-18T:00:00:00, etc.